MWI: Extend max bot TTL to 7 days

timothyb89 opened the pull request:
This change extends the maximum allowed lifetime for bot certificates to 7 days, up from the default maximum value of 24 hours. The default values (1 hour TTL, 20 minute renewal) remain unchanged, and bots will continue to only be able to request 12 hour long certificates due to the default, pre-set 12 hour `max_session_ttl` on the bot roles.

However, users may now explicitly opt in to a longer `max_session_ttl` using a new `max_session_ttl` field in the bot resource, or by setting this with the new `tctl` flag at creation time. For example:
```
$ tctl bots add example --max-session-ttl 168h --roles access
$ tctl get bot/example
kind: bot
metadata:
  name: example
spec:
  max_session_ttl:
    seconds: 604800
  roles:
  - access
status:
  role_name: bot-example
  user_name: bot-example
version: v1
```

For some background, there are 2 clamps on renewable certificate TTLs today:
1. Renewable cert requests are clamped to `libdefaults.MaxRenewableCertTTL`
2. All certs requests are clamped to the lowest `MaxSessionTTL` on any role assigned to the user/bot (by way of `RoleSet.AdjustSessionTTL()`)

   Bot internal roles have traditionally configured a 12 hour `MaxSessionTTL`, and all other roles will default to 30 hours per `apidefaults.MaxCertDuration`. Despite the name, this is just the default value for a role's `MaxSessionTTL`. 

Given that, allowing these longer-lived bots required two changes:
1. Increasing `libdefaults.MaxRenewableCertTTL` to 7 days to pass clamp 1
2. Allowing users to more easily configure the role's `MaxSessionTTL` to pass clamp 2

Existing bots will not be affected by this change and will keep whichever `MaxSessionTTL` they were created with, unless updated manually. Additionally, any roles assigned to bots that apply to output certs will be clamped to the lowest `MaxSessionTTL` in the role set, which is almost certain to be 30 hours or less.

changelog: Machine and Workload Identity: Increase the maximum allowed bot certificate TTL to 7 days, up from 24 hours. Larger values than the default 12 hours must be explicitly requested using the new `--max-session-ttl` flag in `tctl bots add`.
Closes #53380

timothyb89 committed:
MWI: Extend max bot TTL to 7 days

This change extends the maximum allowed lifetime for bot certificates
to 7 days, up from the default maximum value of 24 hours. The default
values (1 hour TTL, 20 minute renewal) remain unchanged.

This somewhat circumvents the cluster's session TTL limit of 30
hours, however this limit is soft and can be overridden via roles
that specify an alternative MaxSessionTTL. This change takes
advantage of this behavior by setting this longer MaxSessionTTL on
the auto-generated bot roles, raising it from the previous default
of 12 hours.

As a result of this implementation detail, existing bots will not be
affected by this change, and will keep whichever MaxSessionTTL they
were created with, unless changed manually. Additionally, this can be
overridden with assigned bot roles as desired.

timothyb89 commented on lib/auth/machineid/machineidv1/bot_service.go:
```diff
@@ -686,7 +687,7 @@ func botToUserAndRole(bot *pb.Bot, now time.Time, createdBy string) (types.User,
 	role, err := types.NewRole(resourceName, types.RoleSpecV6{
 		Options: types.RoleOptions{
-			MaxSessionTTL: types.Duration(12 * time.Hour),
+			MaxSessionTTL: types.Duration(defaults.MaxRenewableCertTTL),
```
Maybe worthy of discussion, a possible alternative to just setting `MaxSessionTTL` to 7 days (which is a bit long) would be to make this an option in the bot proto spec and plumb it through to here. Then users would need to explicitly opt in to the longer max session length on the server for bots to be able to request certs of that length.

I'm already leaning toward that being the better implementation, and if nothing else would probably consider this version as written to be a breaking change that probably should not be backported to v17; I think opt in behavior might be more backport safe.

strideynet replied:
👍  Discussed in slack - sounds like a great plan to me.

timothyb89 replied:
Thanks for the feedback, I've made the change. We're now keeping old behavior with the automatic 12hr limit, but it can be overridden via the bot resource or via a new flag in `tctl bots add` / `tctl bots update --set-max-session-ttl`.

timothyb89 commented on lib/auth/machineid/machineidv1/bot_service.go:
```diff
@@ -686,7 +687,7 @@ func botToUserAndRole(bot *pb.Bot, now time.Time, createdBy string) (types.User,
 	role, err := types.NewRole(resourceName, types.RoleSpecV6{
 		Options: types.RoleOptions{
-			MaxSessionTTL: types.Duration(12 * time.Hour),
+			MaxSessionTTL: types.Duration(defaults.MaxRenewableCertTTL),
```
Another quirk: since all roles have an implicit 30h `max_session_ttl` unless explicitly set, and the resulting limit is the lowest of all assigned roles, output identities will have a practical 30h limit unless every role assigned to the bot also has the limit raised.

The bot's internal identity will only have the main bot role attached, though, so it can request 7d valid certs as desired. The bot warns about this now, e.g.:
```
2025-04-09T19:22:49.253-06:00 WARN [TBOT:SVC:] The server returned an identity shorter than the requested TTL, probably due to a `max_session_ttl` configured on a server-side role. It may not remain valid as long as expected. output:identity-output (directory: ./tbot-user) requested_ttl:168h0m0s renewal_interval:20m0s effective_ttl:30h0m59.746444s expires:2025-04-11T07:22:49.000Z roles:[access] tbot/output_utils.go:439
```

I think this is probably okay since the bot will be recoverable for the full 7d period, and the renewal interval should be under 30h regardless, but curious if others think the UX is acceptable.

timothyb89 committed:
Make max_session_ttl configurable via Bot resource

timothyb89 committed:
Fix tests; fix inclusion of field in generated bot specs; fix update

timothyb89 committed:
Update operator CRDs

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bots-7d-max-session-ttl

timothyb89 committed:
Attempt to fix integration test

boxofrad commented on examples/chart/teleport-cluster/charts/teleport-operator/operator-crds/resources.teleport.dev_botsv1.yaml:
```diff
@@ -35,6 +35,12 @@ spec:
+              max_session_ttl:
+                description: The max session TTL value for the bot's internal role.
+                  Unless specified, bots may not request a value beyond the default
+                  maximum TTL of 12 hours. This value may not be larger than 168 hours.
```
nit: Is there a reason to say "168 hours" rather than "7 days". It just feels slightly less intuitive to me.

strideynet replied:
Yeah - definitely a nit but I'll usually do something like `7 days (168 hours)` for longer time periods in customer facing documentation.

timothyb89 replied:
The thought was to match units, but you're right, that flows a lot more naturally. I've updated the docstring!

strideynet commented:
Do we need to make any changes to tbot's config validation? iirc we throw an error if the configured duration is over 24h.

timothyb89 commented:
> Do we need to make any changes to tbot's config validation? iirc we throw an error if the configured duration is over 24h.

Nope, it checks against the `libdefaults.MaxRenewableCertTTL` constant, so was implicitly updated by this change. It's also just warning level - users can request whatever they like, it'll just warn if something is likely to fail.

timothyb89 committed:
Update docstring with sane units

timothyb89 committed:
Merge branch 'master' into timothyb89/bots-7d-max-session-ttl

timothyb89 committed:
Update crdgen test data

timothyb89 committed:
Merge branch 'timothyb89/bots-7d-max-session-ttl' of github.com:gravitational/teleport into timothyb89/bots-7d-max-session-ttl

timothyb89 committed:
Fix failing tests by enabling UseProtoNames

Also fixes this in the TeleportWorkloadIdentityV1 resource, which was
affected by the same problem.

timothyb89 committed:
Merge branch 'master' into timothyb89/bots-7d-max-session-ttl
